backend/push.js

```diff
@@ -1,53 +1,346 @@
+      return
+    }
+    if (req.status === 413) {
+      throw new Error('Payload too large')
+    }
+  }
+}
 
 export const pushServerActionhandlers: any = {
   [PUSH_SERVER_ACTION_TYPE.SEND_PUBLIC_KEY] () {
     const socket = this
-    socket.send(createMessage(REQUEST_TYPE.PUSH_ACTION, publicKey))
+    socket.send(createMessage(REQUEST_TYPE.PUSH_ACTION, { type: PUSH_SERVER_ACTION_TYPE.SEND_PUBLIC_KEY, data: getVapidPublicKey() }))
   },
-  [PUSH_SERVER_ACTION_TYPE.STORE_SUBSCRIPTION] (payload) {
+  async [PUSH_SERVER_ACTION_TYPE.STORE_SUBSCRIPTION] (payload) {
     const socket = this
-    const subscription = JSON.parse(payload)
+    const { server } = socket
+    const subscription = payload
+    const subscriptionId = await getSubscriptionId(subscription)
 
-    // Reference: is it safe to use 'endpoint' as a unique identifier of a push subscription
-    // (https://stackoverflow.com/questions/63767889/is-it-safe-to-use-the-p256dh-or-endpoint-keys-values-of-the-push-notificatio)
-    socket.server.pushSubscriptions[subscription.endpoint] = subscription
+    if (!server.pushSubscriptions[subscriptionId]) {
+      // If this is a new subscription, we call `subscriptionInfoWrapper` and
+      // store it in memory.
+      server.pushSubscriptions[subscriptionId] = subscriptionInfoWrapper(subscriptionId, subscription)
+      addSubscriptionToIndex(subscriptionId).then(() => {
+        return sbp('chelonia/db/set', `_private_webpush_${subscriptionId}`, JSON.stringify({ subscription: subscription, channelIDs: [] }))
+      }).catch((e) => console.error(e, 'Error saving subscription', subscriptionId))
+      // Send an initial push notification to verify that the endpoint works
+      // This is mostly for testing to be able to auto-remove invalid or expired
+      // endpoints. This doesn't need more error handling than any other failed
+      // call to `postEvent`.
+      postEvent(server.pushSubscriptions[subscriptionId], JSON.stringify({ type: 'initial' })).catch(e => console.warn(e, 'Error sending initial push notification'))
+    } else {
+      // Otherwise, if this is an _existing_ push subscription, we don't need
+      // to call `subscriptionInfoWrapper` but we need to stop sending messages
+      // over the push subscription (since we now have a WS to use, the one
+      // over which the message came).
+      // We expect `server.pushSubscriptions[subscriptionId].sockets.size` to
+      // be `0` when the WS connection has been closed and has since reconnected
+      // If it's not 0, we've already run this code at least once and we don't
+      // need to run it again.
+      if (server.pushSubscriptions[subscriptionId].sockets.size === 0) {
+        server.pushSubscriptions[subscriptionId].subscriptions.forEach((channelID) => {
+          if (!server.subscribersByChannelID[channelID]) return
+          server.subscribersByChannelID[channelID].delete(server.pushSubscriptions[subscriptionId])
+        })
+      }
```

Comment:
Two issues:

1. This code is almost identical to the snippet in `close ()` in `server.js`, could you please DRY it?
2. I noticed that the one difference between the two sections is that the block in `server.js` does `server.subscribersByChannelID[channelID] = new Set()`. I checked to see where this is done now here, and noticed that it seems to be in a now-commented section below. By DRYing it we'll also avoid this type of bug.

Reply:
I'm not sure what you mean. There is a (strong) structural resemblance (if you consider the outer `if` and the `forEach`), but the two pieces of code are fundamentally different in function. The one difference you noticed the `new Set` part, shouldn't be done here (it wouldn't be necessarily incorrect, but it'd be creating resources that aren't needed).

Crucially, however, the difference is in line `294`. Notice that that has **delete**, whereas the corresponding line in `server.js`, `253` uses **add**. Which is to say, I'm not sure this is a DRY violation. If I were reviewing my own code, I might have said to make both blocks into separate functions and call those instead to reduce some of the complexity in a single function, but I wouldn't call this unnecessary repetition.

Reply:
More on the bug you pointed at: The `new Set` _is_ needed in `server.js` and in the now commented code because the set needs to exist to be able to add to it. However, if we're only deleting elements from it, we don't need to create an empty set first, we can just return, as the function does.